cluster-autoscaler/cloudprovider/gce/autoscaling_gce_client.go

```diff
@@ -0,0 +1,194 @@
+		if op, err := client.gceService.ZoneOperations.Get(project, zone, operation.Name).Do(); err == nil {
+			glog.V(4).Infof("Operation %s %s %s status: %s", project, zone, operation.Name, op.Status)
+			if op.Status == "DONE" {
+				return nil
+			}
+		} else {
+			glog.Warningf("Error while getting operation %s on %s: %v", operation.Name, operation.TargetLink, err)
+		}
+	}
+	return fmt.Errorf("Timeout while waiting for operation %s on %s to complete.", operation.Name, operation.TargetLink)
+}
+
+func (client *autoscalingGceClientV1) DeleteInstances(migRef GceRef, instances []*GceRef) error {
+	req := gce.InstanceGroupManagersDeleteInstancesRequest{
+		Instances: []string{},
+	}
+	for _, i := range instances {
+		req.Instances = append(req.Instances, GenerateInstanceUrl(*i))
+	}
+	op, err := client.gceService.InstanceGroupManagers.DeleteInstances(migRef.Project, migRef.Zone, migRef.Name, &req).Do()
+	if err != nil {
+		return err
+	}
+	return client.waitForOp(op, migRef.Project, migRef.Zone)
+}
+
+func (client *autoscalingGceClientV1) FetchMigInstances(migRef GceRef) ([]GceRef, error) {
+	instances, err := client.gceService.InstanceGroupManagers.ListManagedInstances(migRef.Project, migRef.Zone, migRef.Name).Do()
+	if err != nil {
+		glog.V(4).Infof("Failed MIG info request for %s %s %s: %v", migRef.Project, migRef.Zone, migRef.Name, err)
+		return nil, err
+	}
+	refs := []GceRef{}
+	for _, i := range instances.ManagedInstances {
+		ref, err := ParseInstanceUrlRef(i.Instance)
+		if err != nil {
+			return nil, err
+		}
+		refs = append(refs, ref)
+	}
+	return refs, nil
+}
+
+func (client *autoscalingGceClientV1) FetchZones(region string) ([]string, error) {
+	r, err := client.gceService.Regions.Get(client.projectId, region).Do()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get zones for GCE region %s: %v", region, err)
+	}
+	zones := make([]string, len(r.Zones))
+	for i, link := range r.Zones {
+		zones[i] = path.Base(link)
+	}
+	return zones, nil
+}
+
+func (client *autoscalingGceClientV1) FetchMigTemplate(migRef GceRef) (*gce.InstanceTemplate, error) {
+	igm, err := client.gceService.InstanceGroupManagers.Get(migRef.Project, migRef.Zone, migRef.Name).Do()
+	if err != nil {
+		return nil, err
+	}
+	templateUrl, err := url.Parse(igm.InstanceTemplate)
+	if err != nil {
+		return nil, err
+	}
+	_, templateName := path.Split(templateUrl.EscapedPath())
+	return client.gceService.InstanceTemplates.Get(migRef.Project, templateName).Do()
+}
+
+func (client *autoscalingGceClientV1) FetchMigs(zone string, name *regexp.Regexp) ([]string, error) {
+	filter := fmt.Sprintf("name eq %s", name)
+	links := make([]string, 0)
+	req := client.gceService.InstanceGroups.List(client.projectId, zone).Filter(filter)
+	if err := req.Pages(context.TODO(), func(page *gce.InstanceGroupList) error {
+		for _, ig := range page.Items {
+			links = append(links, ig.SelfLink)
+			glog.V(3).Infof("autodiscovered managed instance group %s using regexp %s", ig.Name, name)
```

Comment:
This is only used for autodiscovery, right? Anyway, if this is just a wrapper on GCE client it shouldn't really know about autodiscovery. Can we just change the log to "found managed instance group %s matching regexp %s"?

Reply:
Done